core/vector.go

```diff
@@ -0,0 +1,20 @@
+package core
+
+// Vector is an ordered collection providing fast random access.
+type Vector interface {
+	Seqable
+	SExpressable
```

Comment:
I feel, may be we should make the `Seqable` and `SExpressable` optional? For both features we have a sensible fallback.

For `Seqable`: 

```go
func Seq(v core.Any) (Seq, error) {
   switch s := v.(type) {
      case Seqable:
            return s.Seq()
      case Vector:
            // iterate and create Seq
      default:
            return nil, errors.New("don't know how to create Seq from ??")
    }
}
```

Also, I have a general concern about `SExpressable`. We allow users to change everything about the reader. Including syntax of a `Vector` for example (i.e., I can use `<1 2 3>` to denote a vector in my usage of Slurp). But the value types implementing `SExpressable` use some fixed formatting (e.g., `[1 2 3]` in vector) which now creates a mismatch in what was read as a vector vs what s-expr is emitted.. It's probably not a big issue since usually syntax is fairly similar across LISPs.

Reply:
>I feel, may be we should make the Seqable and SExpressable optional? For both features we have a sensible fallback.

👍 For `SExpressable`. 

I hesitate a _little_  bit to remove `Sequable` from `Vector`,  because a vector is partly defined by the fact that it is a sequence type.  I can't think of a situation where a non-iterable vector is desirable.

Do you have a specific case in mind where forcing `Vector` to be a `Sequable` presents a problem?

> Also, I have a general concern about SExpressable.

This is a good point.  Here are possibilities I currently see:

1. Define some sort of symbol table for use by `repl.Printer`.  This would effectively be the inverse function to the macro table in `Reader`.  Implementation details are a bit unclear to me at this stage.
2. Define some kind of parametrizable factory that knows about delimiters for a given type.
  Think: `NewVectorFactory("[", "]")`.

I'm going out on a limb here, but I think option 2 has the potential to solve a longstanding annoyance of mine in this project:  the fact that the built-in reader macros are tightly coupled to the built-in type implementations.  I would really like to be able to substitute e.g. my own `String` implementation without having to re-implement `readString`.  Here's what that might look like:

```go
package reader

func NewKeywordReader(newString func(string) (core.Any, error)) Macro {
        func (rd *Reader, init rune) (core.Any, error) {
        	beginPos := rd.Position()

	        token, err := rd.Token(-1)
        	if err != nil {
		        return nil, rd.annotateErr(err, beginPos, token)
	        }

        	return newString(s)
        }
}

func NewListReader(newList func(...core.Any) (core.Any, error)) Macro {
        func (rd *Reader, init rune) (core.Any, error) {
		const listEnd = ')'

		beginPos := rd.Position()

		forms := make([]core.Any, 0, 32) // pre-allocate to improve performance on small lists
		if err := rd.Container(listEnd, "list", func(val core.Any) error {
			forms = append(forms, val)
			return nil
		}); err != nil {
			return nil, rd.annotateErr(err, beginPos, "")
		}

		return newList(forms...)
        }
}
```

The macro table could then be assembled in the following way:

```go
	macros: map[rune]Macro{
	'"':  NewStringReader(NewStringFactory("\"", "\"")),
	// ...
	'(':  NewListReader(NewListFactory("(", ")")),
	')':  UnmatchedDelimiter(),
	// ...
},
```

Under the hood `NewStringFactory` would parametrize the `String` type to use double-quotes, and the `List` type to use parens.

Thoughts?  (I haven't thought of this very hard, so I might be overlooking some obvious issues...)

Reply:
>I hesitate a little bit to remove Sequable from Vector, because a vector is partly defined by the fact that it is a sequence type. I can't think of a situation where a non-iterable vector is desirable.

Nevermind, there was some confusion on my end as to the meaning of `Seq`.  I was thinking of `Seq` as an analog to `Iterable`, but it's actually more like `ListSemanticsProvider`.

Mental model updated. ⚡ 🧠 

(Also, 👍 for removing `Sequable` from `Vector` interface.)

Reply:
Seq is still a sequence type. And vector is definitely a sequence. but by embedding seqable we are saying "vector should be convertible to a sequence" and not "vector itself is a sequence" (this is fine since this is our requirement anyway)

Since our requirement is that we should be able to convert a vector to a sequence type, and considering keeping things easy and simple is always a goal, I suggested the ToSeq approach. How it helps?

1. Implementing a custom vector type is easier (since I don't need to implement the seqable part if I don't want to)
2. I can make the vector itself a seq (instead of seqable) if there is a optimal way to look at my vector implementation as a seq. (This might be problematic too since we expect to be able to differentiate vector and a list 😅 so may be wishful thinking)
3. I can make my vector Seqable if I want to optimise it. 

ToSeq will take care of choosing the most appropriate variant and converting it to a Seq. 

Also, we could add logic to convert Go native arrays and slices to a Seq in ToSeq.

Reply:
Regarding the reader/printer, I'm also not sure how to generalize it. But the simplest approach I can think of is to let the user build a printer which relies on type switch to emit lisp source based on their reader setting. 

(Repl package could have one such function based on the built-in types and the reader config repl uses).

Reply:
Re Sequable:  Thanks for the clarifications.  👍 On my end.

Re SExpressable:  should we just shelf this discussion for now and come back to it later?

Reply:
Yep. I think that makes sense. Print is not priority.. also I don't think I will get any time to go through this soon. feel free to merge when you address these comments

Reply:
>also I don't think I will get any time to go through this soon [...] feel free to merge when you address these comments.

Will do.   I hope everything is going well on your end. 🙂 

Thanks for the feedback!